src/greenpak/i2c.py

```diff
@@ -133,72 +174,112 @@ def __empty_rd(self, addr: int):
 
     @override
     def write(self, addr: int, data: bytearray, silent: bool = False) -> bool:
+        raise NotImplementedError("SMBus wrapper doesn't support it")
+
+    @override
+    def read(
+        self, addr: int, byte_count: int, silent: bool = False
+    ) -> bytearray | None:
+        raise NotImplementedError("SMBus wrapper doesn't support it")
+
+    @override
+    def gp_read(self, i2c_addr: int, start: int, byte_count: int) -> bytearray | None:
+        try:
+            if 0 == byte_count:
+                # assume caller probing for GP devs on bus
+                self.__empty_rd(i2c_addr)  # Can/will throw if no dev found.
+                return bytearray()
+            else:
+                offset = start
+                data = []
+                while byte_count > 0:
+                    xferlen = min(self.smbus2.I2C_SMBUS_BLOCK_MAX, byte_count)
+                    data.extend(self.bus.read_i2c_block_data(i2c_addr, offset, xferlen))
+                    offset += xferlen
+                    byte_count -= xferlen
+                return bytearray(data)
+        except Exception as e:
+                if self.trace_errors == True:
+                    str = "SMBus: while reading i2c addr 0x%02x: caught exception: %s" % (
+                        i2c_addr,
+                        e
+                    )
+                    print(str)
+        return None
+
+    @override
+    def gp_write(self, i2c_addr: int, start: int, data: bytearray) -> bool:
         try:
+            # Put them here so can trace in case of an exception (see below)
+            regaddr   = start
+            bytesleft = len(data)
+            offset    = 0
+
             if 0 == len(data):
                 # special case as we not actually writing/sending data.
                 # (assume caller probing for GP devs on bus).
-                self.__empty_wr(addr)  # Can throw
-            elif 2 == len(data):
+                self.__empty_wr(i2c_addr)  # Can throw
+            elif 1 == len(data):
```

Comment:
Please refactor this special case to a separate method and call it if the condition of the special case is met. 

Condition is 
1. len==1, 
2. start = e3. 
3. The memory space bits in the i2c address are REGISTER space.

Reply:
ok, 
note that 3rd condition is only implied, the smbus adapter doesn't know of different memory spaces, reg nmv or eep (has no knowledge of them).

we can refactor to method based on first 2, and that 3rd one is special case checked if error happned, as per current handling of it.